Source/Core/Core/PowerPC/Tables/TableGen.cpp

```diff
@@ -0,0 +1,476 @@
+  std::bitset<64> subtables;
+  size_t instruction_offset = 0;
+  size_t subtable_offset = 0;
+  int shift = 0;
+  int length = 0;
+  DecodingEntry(std::string name) : table_name(std::move(name)) {}
+};
+
+static void CreateTable(std::vector<Instruction>& table, std::vector<DecodingEntry>& decoding_table,
+                        DecodingEntry& entry, const std::vector<InputLine>& lines)
+{
+  entry.instruction_offset = table.size();
+  entry.subtable_offset = decoding_table.size();
+  std::vector<size_t> subtables;
+  size_t start = 0;
+  while (true)
+  {
+    if (start == lines.size())
+    {
+      std::cerr << "Error: subtable \"" << entry.table_name << "\" not found\n";
+      std::exit(1);
+    }
+    if (std::holds_alternative<StartTableMarker>(lines[start]))
+    {
+      auto& marker = std::get<StartTableMarker>(lines[start]);
+      if (marker.name == entry.table_name)
+      {
+        start += 1;
+        entry.description = marker.description;
+        entry.shift = marker.shift;
+        entry.length = marker.length;
+        break;
+      }
+    }
+    start += 1;
+  }
+  size_t end;
+  for (end = start; end < lines.size() && end - start < (static_cast<size_t>(1) << entry.length) &&
+                    !std::holds_alternative<StartTableMarker>(lines[end]);
+       end += 1)
+  {
+    entry.instructions <<= 1;
+    entry.subtables <<= 1;
+    if (std::holds_alternative<Instruction>(lines[end]))
+    {
+      table.push_back(std::get<Instruction>(lines[end]));
+      entry.instructions |= 1;
+    }
+    else if (std::holds_alternative<SubTable>(lines[end]))
+    {
+      subtables.push_back(end);
+      entry.subtables |= 1;
+    }
+  }
+  if (end - start == (static_cast<size_t>(1) << entry.length))
+  {
+    for (size_t i = end; i < lines.size() && !std::holds_alternative<StartTableMarker>(lines[i]);
+         i += 1)
+    {
+      if (!std::holds_alternative<Empty>(lines[i]))
+      {
+        std::cerr << "Error: subtable \"" << entry.table_name << "\" (" << entry.length
+                  << "-bit field) is longer than " << (1 << entry.length) << "\n";
+        std::exit(1);
+      }
+    }
+  }
+  entry.instructions <<= 64 + start - end;
+  entry.subtables <<= 64 + start - end;
+  const int offset = entry.subtable_offset;
+  for (auto& i : subtables)
+  {
+    // WARNING: may invalidate 'entry'
+    decoding_table.push_back(DecodingEntry(std::get<SubTable>(lines[i]).name));
+  }
+  for (size_t i = 0; i < subtables.size(); i += 1)
+  {
+    CreateTable(table, decoding_table, decoding_table[offset + i], lines);
+  }
+}
+
+static void DoOutput(const std::vector<Instruction>& table,
+                     const std::vector<DecodingEntry>& decoding_table, const OutputOptions& options,
+                     std::ostream& out)
+{
+  if ((options.flags & OPID_RANGES) || (options.flags & DECODING_TABLE) || !options.columns.empty())
+  {
+    out << "// Auto-generated by TableGen.cpp\n// clang-format off\n";
```

Comment:
A line of the file should match the regex of `^// Code generated .* DO NOT EDIT\.$`

suggested: `// Code generated by TableGen.cpp; DO NOT EDIT.\n// clang-format off\n`